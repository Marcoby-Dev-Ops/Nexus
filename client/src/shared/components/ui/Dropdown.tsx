import React, { useState, useRef, useEffect } from 'react';
import { cn } from '@/shared/lib/utils';

export interface DropdownItem {
  label: string;
  onClick?: () => void;
  disabled?: boolean;
}

export interface DropdownProps {
  trigger: React.ReactNode;
  items: DropdownItem[];
  className?: string;
  align?: 'left' | 'right';
}

/**
 * Simple dropdown component for 1.0
 */
const Dropdown: React.FC<DropdownProps> = ({ 
  trigger, 
  items, 
  className, 
  align = 'left' 
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
        setIsOpen(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, []);

  const handleItemClick = (item: DropdownItem) => {
    if (!item.disabled && item.onClick) {
      item.onClick();
    }
    setIsOpen(false);
  };

  return (
    <div className={cn('relative inline-block', className)} ref={dropdownRef}>
      <div onClick={() => setIsOpen(!isOpen)} className="cursor-pointer">
        {trigger}
      </div>
      
      {isOpen && (
        <div 
          className={cn(
            'absolute z-dropdown mt-1 w-56 rounded-md border border-border bg-popover text-popover-foreground shadow-lg',
            'ring-1 ring-black/5 dark:ring-white/10',
            'animate-in fade-in-0 zoom-in-95 duration-200',
            align === 'right' ? 'right-0' : 'left-0'
          )}
        >
          <div className="p-1">
            {items.map((item, index) => (
              <button
                key={index}
                onClick={() => handleItemClick(item)}
                disabled={item.disabled}
                className={cn(
                  'block w-full px-2 py-2 text-left text-sm rounded-sm transition-colors duration-150',
                  'hover:bg-accent hover:text-accent-foreground focus:bg-accent focus:text-accent-foreground',
                  'outline-none',
                  item.disabled
                    ? 'text-muted-foreground cursor-not-allowed opacity-50'
                    : 'text-popover-foreground cursor-pointer'
                )}
              >
                {item.label}
              </button>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default Dropdown; 
