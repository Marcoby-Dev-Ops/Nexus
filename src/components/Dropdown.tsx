import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';

/**
 * @interface DropdownProps
 * @description Props for the Dropdown component.
 */
export interface DropdownProps {
  label: React.ReactNode;
  children: React.ReactNode;
  className?: string;
}

/**
 * @name Dropdown
 * @description A dropdown menu component for actions and navigation.
 * @param {DropdownProps} props - The props for the component.
 * @returns {JSX.Element} The rendered Dropdown component.
 */
export const Dropdown: React.FC<DropdownProps> = React.memo(({ label, children, className = '' }) => {
  const [open, setOpen] = useState(false);
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    function handleClickOutside(event: MouseEvent) {
      if (ref.current && !ref.current.contains(event.target as Node)) {
        setOpen(false);
      }
    }
    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, []);

  return (
    <div className={`relative inline-block text-left ${className}`} ref={ref}>
      <button onClick={() => setOpen((o) => !o)} className="inline-flex justify-center w-full rounded-md border border-input bg-background px-4 py-2 text-sm font-medium text-foreground shadow-sm hover:bg-muted focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2">
        {label}
        <svg className="ml-2 h-4 w-4" viewBox="0 0 20 20" fill="currentColor"><path fillRule="evenodd" d="M5.23 7.21a.75.75 0 011.06.02L10 10.585l3.71-3.355a.75.75 0 111.02 1.1l-4.25 3.85a.75.75 0 01-1.02 0l-4.25-3.85a.75.75 0 01.02-1.06z" clipRule="evenodd" /></svg>
      </button>
      {open && (
        <div className="absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-popover shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none">
          <div className="py-1">{children}</div>
        </div>
      )}
    </div>
  );
});

Dropdown.displayName = 'Dropdown';

Dropdown.propTypes = {
  label: PropTypes.node.isRequired,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
};

export default Dropdown; 