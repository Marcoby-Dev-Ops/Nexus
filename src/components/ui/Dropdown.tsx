import React, { useState, useRef, useEffect } from 'react';
import { cn } from '@/lib/utils';

export interface DropdownItem {
  label: string;
  value: string;
  disabled?: boolean;
  onClick?: () => void;
}

export interface DropdownProps {
  trigger: React.ReactNode;
  items: DropdownItem[];
  className?: string;
  align?: 'left' | 'right';
}

/**
 * Simple dropdown component for 1.0
 */
const Dropdown: React.FC<DropdownProps> = ({ 
  trigger, 
  items, 
  className, 
  align = 'left' 
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
        setIsOpen(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, []);

  const handleItemClick = (item: DropdownItem) => {
    if (!item.disabled && item.onClick) {
      item.onClick();
    }
    setIsOpen(false);
  };

  return (
    <div className={cn('relative inline-block', className)} ref={dropdownRef}>
      <div onClick={() => setIsOpen(!isOpen)} className="cursor-pointer">
        {trigger}
      </div>
      
      {isOpen && (
        <div 
          className={cn(
            'absolute z-10 mt-1 w-56 rounded-md bg-card shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none',
            align === 'right' ? 'right-0' : 'left-0'
          )}
        >
          <div className="py-1">
            {items.map((item, index) => (
              <button
                key={index}
                onClick={() => handleItemClick(item)}
                disabled={item.disabled}
                className={cn(
                  'block w-full px-4 py-2 text-left text-sm',
                  item.disabled
                    ? 'text-muted-foreground cursor-not-allowed'
                    : 'text-foreground/90 hover:bg-muted hover:text-foreground'
                )}
              >
                {item.label}
              </button>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default Dropdown; 