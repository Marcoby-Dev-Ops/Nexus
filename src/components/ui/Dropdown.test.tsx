/**
 * @file Dropdown.test.tsx
 * @description Unit and snapshot tests for the Dropdown component.
 */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Dropdown from './Dropdown';

/**
 * Mock props for Dropdown
 */
const options = [
  { label: 'Option 1', value: '1' },
  { label: 'Option 2', value: '2' },
];

describe('Dropdown', () => {
  it('renders with options', () => {
    render(<Dropdown options={options} onChange={jest.fn()} />);
    expect(screen.getByRole('combobox')).toBeInTheDocument();
    expect(screen.getByText('Option 1')).toBeInTheDocument();
    expect(screen.getByText('Option 2')).toBeInTheDocument();
  });

  it('calls onChange when an option is selected', () => {
    const handleChange = jest.fn();
    render(<Dropdown options={options} onChange={handleChange} />);
    fireEvent.change(screen.getByRole('combobox'), { target: { value: '2' } });
    expect(handleChange).toHaveBeenCalledWith('2');
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Dropdown options={options} onChange={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });
}); 