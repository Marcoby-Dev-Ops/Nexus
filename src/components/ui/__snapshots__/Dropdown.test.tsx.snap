// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative inline-block"
  >
    <div
      class="cursor-pointer"
    >
      <button
        class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
      >
        Open
      </button>
    </div>
  </div>
</DocumentFragment>
`;
